src/algebra/quaternion.lean

```diff
@@ -133,9 +133,9 @@ by refine_struct
     sub := has_sub.sub,
     mul := (*),
     one := 1,
-    nsmul := @nsmul_rec _ ⟨0⟩ ⟨(+)⟩,
-    gsmul := @gsmul_rec _ ⟨0⟩ ⟨(+)⟩ ⟨has_neg.neg⟩,
-    npow := @npow_rec _ ⟨1⟩ ⟨(*)⟩ };
+    nsmul := @nsmul_rec _ ⟨(0 : ℍ[R, c₁, c₂])⟩ ⟨(+)⟩,
+    gsmul := @gsmul_rec _ ⟨(0 : ℍ[R, c₁, c₂])⟩ ⟨(+)⟩ ⟨has_neg.neg⟩,
+    npow := @npow_rec _ ⟨(1 : ℍ[R, c₁, c₂])⟩ ⟨(*)⟩ };
```

Comment:
I don't get why this is necessary; surely there's no coe_sort or coe_fn involved here?

Reply:
This is an elaboration order issue I think: the `0` gets elaborated too early, so its type is defaulted to `ℕ`, which later causes a type error because `nsmul_rec` should have type `ℕ → ℍ[R, c₁, c₂] → ℍ[R, c₁, c₂]` instead of `ℕ → ℕ → ℕ`.

Reply:
I don't understand why the elaboration behavior changed though on expressions with no coercions

Reply:
There were some elaboration issues with new coercions. They were fixed by @gebner. It seems that these fixes had side effects.

Reply:
It would be good to confirm these side effects are an expected compromise rather than an accident.